File: packages/dotnet/OpenSkp/Parser.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace OpenSkp
{
    /// <summary>High-level entry point for opening and parsing .skp files.
    ///
    /// <code>
    /// SkpModel model = SkpFile.Open("house.skp");
    /// Console.WriteLine(model.Version);
    /// foreach (var layer in model.Layers) Console.WriteLine(layer.Name);
    /// </code>
    /// </summary>
    public static class SkpFile
    {
        public static SkpModel Parse(byte[] buffer)
        {
            Core.RawParsed parsed;
            try
            {
                parsed = Core.FullParse(buffer);
            }
            catch (LegacyParseError e)
            {
                throw new ArgumentException($"legacy .skp parse failed: {e.Message}", e);
            }

            var model = new SkpModel { Version = parsed.Version };

            foreach (var kv in parsed.DefsDict)
            {
                model.Definitions[kv.Key] = BuildDefinition(kv.Key, kv.Value);
            }
            model.Root = BuildDefinition(0, parsed.Root);

            foreach (var kv in parsed.LayerColors)
            {
                model.Layers.Add(new Layer { Name = kv.Key, ColorR = kv.Value.R, ColorG = kv.Value.G, ColorB = kv.Value.B });
            }

            var matForData = new Dictionary<Geometry.RawMaterial, Material>(ReferenceEqualityComparer.Instance);
            foreach (var rawMat in parsed.Materials.Values)
            {
                Texture? texture = null;
                if (rawMat.Texture != null)
                {
                    texture = new Texture
                    {
                        Filename = rawMat.Texture.Filename,
                        Width = rawMat.Texture.XScale,
                        Height = rawMat.Texture.YScale,
                        Data = rawMat.Texture.Data,
                    };
                }
                var mat = new Material
                {
                    Name = rawMat.Name,
                    Color = (rawMat.R, rawMat.G, rawMat.B, 255),
                    Transparency = rawMat.Transparency,
                    Texture = texture,
                    Colorized = rawMat.Colorized,
                    ColorizeType = rawMat.ColorizeType,
                };
                model.Materials.Add(mat);
                matForData[rawMat] = mat;
            }

            foreach (var kv in parsed.MaterialIdToName)
            {
                long mId = kv.Key;
                string mName = kv.Value;
                Geometry.RawMaterial? rawMat = parsed.Materials.TryGetValue(mName, out var m1) ? m1
                    : parsed.MaterialsByFolder.TryGetValue(mName, out var m2) ? m2 : null;
                if (rawMat == null || !matForData.TryGetValue(rawMat, out var mat)) continue;
                if (mat.Id == null) mat.Id = mId;
                model.MaterialsById[mId] = mat;
            }

            foreach (var st in parsed.Styles)
            {
                model.Styles.Add(new Style { Name = st.Name, FrontColor = st.FrontColor, BackColor = st.BackColor });
            }

            return model;
        }

        private static Definition BuildDefinition(long defId, Geometry.RawDefinition d)
        {
            var defn = new Definition
            {
                Id = defId,
                Guid = d.Guid ?? "",
                Name = d.Name ?? "",
                AlwaysFacesCamera = d.AlwaysFacesCamera,
                IsImage = d.IsImage,
            };

            foreach (var kv in d.Builder.Vertices)
            {
                var (x, y, z) = kv.Value;
                defn.Vertices[kv.Key] = new Vertex { Id = kv.Key, X = x, Y = y, Z = z };
            }

            foreach (var kv in d.Builder.Edges)
            {
                var (v1, v2) = kv.Value;
                int flags = d.Builder.EdgeFlags.TryGetValue(kv.Key, out var f) ? f : 0;
                defn.Edges[kv.Key] = new Edge
                {
                    Id = kv.Key,
                    V1Id = v1 ?? 0,
                    V2Id = v2 ?? 0,
                    Soft = (flags & 0x08) != 0,
                    Smooth = (flags & 0x10) != 0,
                    Hidden = (flags & 0x01) != 0,
                };
            }

            foreach (var kv in d.Builder.Faces)
            {
                var f = kv.Value;
                defn.Faces[kv.Key] = new Face
                {
                    Id = kv.Key,
                    Loops = f.Loops,
                    Normal = f.Normal,
                    MaterialId = f.MaterialId,
                    BackMaterialId = f.BackMaterialId,
                    UvTransform = f.UvTransform,
                    UvTransformBack = f.UvTransformBack,
                };
            }

            foreach (var inst in d.Builder.Instances)
            {
                defn.Instances.Add(new Instance
                {
                    Name = inst.Name ?? "",
                    RefIdx = inst.RefIdx,
                    Guid = inst.RefGuid ?? "",
                    Matrix = inst.Matrix,
                    MaterialId = inst.MaterialId,
                });
            }

            return defn;
        }

        public static SkpModel Open(string filePath)
        {
            var p = Path.GetFullPath(filePath);
            if (!File.Exists(p))
            {
                throw new FileNotFoundException($"File not found: {p}");
            }
            if (!string.Equals(Path.GetExtension(p), ".skp", StringComparison.OrdinalIgnoreCase))
            {
                throw new ArgumentException($"Expected a .skp file, got: {Path.GetExtension(p)}");
            }
            var bytes = File.ReadAllBytes(p);
            return Parse(bytes);
        }
    }

    internal sealed class ReferenceEqualityComparer : IEqualityComparer<Geometry.RawMaterial>
    {
        public static readonly ReferenceEqualityComparer Instance = new ReferenceEqualityComparer();
        public bool Equals(Geometry.RawMaterial? x, Geometry.RawMaterial? y) => ReferenceEquals(x, y);
        public int GetHashCode(Geometry.RawMaterial obj) => System.Runtime.CompilerServices.RuntimeHelpers.GetHashCode(obj);
    }
}

File: packages/dart/lib/src/parser.dart
import 'dart:io';
import 'dart:typed_data';

import 'core.dart';
import 'geometry.dart';
import 'legacy.dart';
import 'model.dart';

/// High-level entry point for opening and parsing .skp files.
///
/// ```dart
/// final model = SkpFile.open('house.skp').parse();
/// print(model.version);
/// for (final layer in model.layers) print(layer.name);
/// ```
class SkpFile {
  final String path;
  final Uint8List? _bytes;

  SkpFile._(this.path, this._bytes);

  /// Open a SketchUp file for parsing.
  factory SkpFile.open(String filepath) {
    final f = File(filepath);
    if (!f.existsSync()) {
      throw FileSystemException('File not found', filepath);
    }
    if (!filepath.toLowerCase().endsWith('.skp')) {
      throw ArgumentError('Expected a .skp file, got: $filepath');
    }
    return SkpFile._(filepath, null);
  }

  /// Parse directly from an in-memory buffer (no file I/O).
  factory SkpFile.fromBuffer(Uint8List bytes) {
    return SkpFile._('<memory>', bytes);
  }

  SkpModel parse() {
    final bytes = _bytes ?? File(path).readAsBytesSync();

    RawParsed parsed;
    try {
      parsed = Core.fullParse(bytes);
    } on LegacyParseError catch (e) {
      throw ArgumentError('legacy .skp parse failed: ${e.message}');
    }

    final model = SkpModel()..version = parsed.version;

    for (final entry in parsed.defsDict.entries) {
      model.definitions[entry.key] = _buildDefinition(entry.key, entry.value);
    }
    model.root = _buildDefinition(0, parsed.root);

    for (final entry in parsed.layerColors.entries) {
      final (r, g, b) = entry.value;
      model.layers.add(Layer(name: entry.key, colorR: r, colorG: g, colorB: b));
    }

    final matForData = <RawMaterial, Material>{};
    for (final rawMat in parsed.materials.values) {
      Texture? texture;
      final rawTex = rawMat.texture;
      if (rawTex != null) {
        texture = Texture(
            filename: rawTex.filename,
            width: rawTex.xScale,
            height: rawTex.yScale,
            data: rawTex.data);
      }
      final mat = Material(
        name: rawMat.name,
        color: (rawMat.r, rawMat.g, rawMat.b, 255),
        transparency: rawMat.transparency,
        texture: texture,
        colorized: rawMat.colorized,
        colorizeType: rawMat.colorizeType,
      );
      model.materials.add(mat);
      matForData[rawMat] = mat;
    }

    for (final entry in parsed.materialIdToName.entries) {
      final mId = entry.key;
      final mName = entry.value;
      final rawMat = parsed.materials[mName] ?? parsed.materialsByFolder[mName];
      if (rawMat == null) continue;
      final mat = matForData[rawMat];
      if (mat == null) continue;
      mat.id ??= mId;
      model.materialsById[mId] = mat;
    }

    for (final st in parsed.styles) {
      model.styles.add(Style(
          name: st.name, frontColor: st.frontColor, backColor: st.backColor));
    }

    return model;
  }

  static Definition _buildDefinition(int defId, RawDefinition d) {
    final defn = Definition(
      id: defId,
      guid: d.guid ?? '',
      name: d.name ?? '',
      alwaysFacesCamera: d.alwaysFacesCamera,
      isImage: d.isImage,
    );

    for (final entry in d.builder.vertices.entries) {
      final (x, y, z) = entry.value;
      defn.vertices[entry.key] = Vertex(id: entry.key, x: x, y: y, z: z);
    }

    for (final entry in d.builder.edges.entries) {
      final (v1, v2) = entry.value;
      final flags = d.builder.edgeFlags[entry.key] ?? 0;
      defn.edges[entry.key] = Edge(
        id: entry.key,
        v1Id: v1 ?? 0,
        v2Id: v2 ?? 0,
        soft: (flags & 0x08) != 0,
        smooth: (flags & 0x10) != 0,
        hidden: (flags & 0x01) != 0,
      );
    }

    for (final entry in d.builder.faces.entries) {
      final f = entry.value;
      defn.faces[entry.key] = Face(
        id: entry.key,
        loops: f.loops,
        normal: f.normal,
        materialId: f.materialId,
        backMaterialId: f.backMaterialId,
        uvTransform: f.uvTransform,
        uvTransformBack: f.uvTransformBack,
      );
    }

    for (final inst in d.builder.instances) {
      defn.instances.add(Instance(
        name: inst.name ?? '',
        refIdx: inst.refIdx,
        guid: inst.refGuid ?? '',
        matrix: inst.matrix,
        materialId: inst.materialId,
      ));
    }

    return defn;
  }
}
